src/server/pps/server/api_server.go

```diff
@@ -1065,23 +1066,55 @@ func (a *apiServer) InspectDatum(ctx context.Context, request *pps.InspectDatumR
 func (a *apiServer) ListDatum(request *pps.ListDatumRequest, server pps.API_ListDatumServer) (retErr error) {
 	// TODO: Auth?
 	ensurePipelineProject(request.GetJob().GetPipeline())
+	if request.Job != nil && request.Input != nil {
+		return errors.Errorf("cannot specify both job and input")
+	}
+	var metas []*datum.Meta
 	if request.Input != nil {
-		return a.listDatumInput(server.Context(), request.Input, func(meta *datum.Meta) error {
-			di := convertDatumMetaToInfo(meta, nil)
-			di.State = pps.DatumState_UNKNOWN
-			if !request.Filter.Allow(di) {
-				return nil
-			}
-			return errors.EnsureStack(server.Send(di))
-		})
+		if err := a.listDatumInput(server.Context(), request.Input, func(meta *datum.Meta) error {
+			metas = append(metas, meta)
+			return nil
+		}); err != nil {
+			return errors.EnsureStack(err)
+		}
+	} else {
+		if err := a.collectDatums(server.Context(), request.Job, func(meta *datum.Meta, _ *pfs.File) error {
+			metas = append(metas, meta)
+			return nil
+		}); err != nil {
+			return errors.EnsureStack(err)
+		}
+	}
+	number := request.Number
+	if number == 0 {
+		number = math.MaxInt64
 	}
-	return a.collectDatums(server.Context(), request.Job, func(meta *datum.Meta, _ *pfs.File) error {
+	for i, meta := range metas {
```

Comment:
This is a similar problem to what I mentioned in the list file PR. I think the same strategy I suggested there would work here as well.